src/data/set_like.lean

```diff
@@ -0,0 +1,116 @@
+/-
+Copyright (c) 2021 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Wieser
+-/
+import data.set.basic
+
+/-!
+# Typeclass for a type `A` with an injective map to `set B`
+
+This typecalss is primary for use by subobjects like `submonoid` and `submodule`.
+
+A typical subobject should be declared as:
+```
+structure my_subobject (X : Type*) :=
+(carrier : set X)
+(op_mem : ∀ {x : X}, x ∈ carrier → sorry ∈ carrier)
+
+namespace my_subobject
+
+instance : set_like (my_subobject R) M :=
+⟨sub_mul_action.carrier, λ p q h, by cases p; cases q; congr'⟩
+
+@[simp] lemma mem_carrier {p : my_subobject R} : x ∈ p.carrier ↔ x ∈ (p : set M) := iff.rfl
```

Comment:
I would expect a `simp` lemma that rewrites `p.carrier` to `(p : set M)`, along the lines of `subtype.val_eq_coe`.

Reply:
That's probably a good idea, but I've avoided adding any lemmas to this file that didn't already exist somewhere.
Once we have all the lemmas in one place, it's much easier to have a conversation about redesigning them.